### ISSUE TITLE
First call in the chain?

### ISSUE THREAD

**Issue by joshuawhite929** [joshuawhite929/spmia_feedback#106]
John, in the text, you have the following:

> If the tmx-correlation-id is not present on the HTTP header, this means that the call being made through Zuul is the very first microservice call in the call chain. 

Based on how your examples work, isn't this always true?  Unless a users is passing in a tmx-correlation-id header, you'll always have to generate it.  The part that is confusing for me is the "first microservice call in the call chain" comment.

Based on chapters 5 and 6, your examples work like this:

Request -> Zuul (ribbon/eureka) -> Licensing Service (ribbon/eureka) -> Organization Service

Based on this comment, it sounds like you were inferring that this works like this (which is an alternative approach, but not covered yet)

Request -> Zuul (ribbon/eureka) -> Licensing Service (ribbon/eureka) -> Zuul (ribbon/eureka) -> Organization Service

With the above model, when the licensing service calls the Organization service, it does so by calling Zuul rather than using ribbon/eureka to find and call the service directly.

In the second model, ALL service calls go through the gateway.  I believe in your examples so far, only calls from the outside world go through the gateway.  Inter-service calls are direct.  This challenge with this approach (and perhaps your readers should know) is that the rules that you have in Zuul (think SpecialRoutes filter) won't get executed.

Thoughts?

Note - When you answer this, can you please change the ownership to me so I can know?

**Comment by carnellj** [comment]
Hey Josh,

I reworded the paragraph to try and make it a little clearer.  Here is the new text:

"If the tmx-correlation-id is not present on the HTTP header, our Zuul TrackingFilter will generate and set the correlation ID. If there is already a correlation ID present, Zuul will not do anything with the correlation ID.  The presence of a correlation ID means that this particular service call is part of a chain of service calls carrying out the user’s request. In this case our TrackingFilter class will do nothing.
"

As for your second comments, I am trying to do the book through a progression.  My early examples did not have a service gateway involved so they do call their component services directly.   With chapter 6 I introduce the service gateway and now the invocation of services can really go through 3 different types of models.

1.  Public calls from the outside world calls go through a service gateway and once they are past the gateway, any internal private calls can go directly to one another.
2.  Public calls from the outside world calls go through a service gateway and they once they go through a public service gateway they hit a private service gateway.  The internal private API calls can each other once they get behind the private gateway directly without having to go through the gateway).
3.  The last model (and the one I use throughout the book) says that everything has to go through some type of service gateway even if they are private.  This is the most conservative approach because it guarantees everything goes through a policy enforcement point.

With the company I work at now, we use the second model.  When I worked at a financial services institution, we were conservative to a fault and used number 3.  I kept us using model number in the examples to keep it simple and not have to flip people back and forth.  

I did a section at the end of Chapter 7 were I talked about the concepts above because in many ways the model you choose really depends on the risk you are willing to take around services and security.    Let me know your thoughts.  The challenge I have in Chapter 6 and Chapter 7 is the chapters are on big topics so its extremely difficult to cover everything I want to cover in detail. :-)
